Library/Homebrew/api/cask.rb

```diff
@@ -23,12 +23,20 @@ def fetch_source(token, git_head: nil)
         sig { returns(Hash) }
         def all_casks
           @all_casks ||= begin
-            json_casks = Homebrew::API.fetch_json_api_file "cask.json",
-                                                           target: HOMEBREW_CACHE_API/"cask.json"
+            json_casks, updated = Homebrew::API.fetch_json_api_file "cask.json",
+                                                                    target: HOMEBREW_CACHE_API/"cask.json"
 
-            json_casks.to_h do |json_cask|
+            casks = json_casks.to_h do |json_cask|
               [json_cask["token"], json_cask.except("token")]
             end
+
+            cask_names_path = HOMEBREW_CACHE_API/"parsed/cask_names.txt"
+            if !cask_names_path.exist? || updated
+              (HOMEBREW_CACHE_API/"parsed").mkpath
+              cask_names_path.write(casks.keys.join("\n"))
```

Comment:
How slow is this? It feels like it may be worth doing every time.

Reply:
That has issues when run in sandboxed use cases.

Reply:
@Bo98 What are these use-cases and is this using Homebrew's sandbox or another? 

I guess I'm struggling to understand why this would be a problem if it happens "always" vs. "sometimes" (I can see why "never" would be not a problem).

Reply:
opam is an example of something that runs in a sandbox but may link to Homebrew installed dependencies.

> I guess I'm struggling to understand why this would be a problem if it happens "always" vs. "sometimes" (I can see why "never" would be not a problem).

It was mostly for `HOMEBREW_NO_AUTO_UPDATE`. I don't check for that directly because we do ignore that if the cache doesn't exist at all.

Reply:
> opam is an example of something that runs in a sandbox but may link to Homebrew installed dependencies.

If that's something we want to support we should probably detect it and handle it better. Not sure we should support that, though.

> It was mostly for HOMEBREW_NO_AUTO_UPDATE. I don't check for that directly because we do ignore that if the cache doesn't exist at all.

Where is that set, just by the user/`opam` in this case? Again, feels to me that if there's any chance of a write action happening here then it's something we should handle differently.

Perhaps we just assume all file writing here can fail (e.g. the touch, the download) and handle it accordingly?

Reply:
> If that's something we want to support we should probably detect it and handle it better. Not sure we should support that, though.

There's really no reason why `brew --prefix` (prior to shell implementation) or `brew info --json=v2`, for example, should require writing files to disk (unless the existing data is missing) so it makes sense and seems easy enough to add a test for if you want. It worked fine all the way to 4.0.0.

> Where is that set, just by the user/opam in this case?

Right now, `opam` is using `HOMEBREW_NO_INSTALL_FROM_API`. Opam is just an example though.

Reply:
> Perhaps we just assume all file writing here can fail (e.g. the touch, the download) and handle it accordingly?

Code seems like it might be messy that way but might make sense to do anyway.

Reply:
> should require writing files to disk (unless the existing data is missing)

This is my point. Given that sometimes it will have to write to disk if this data is missing: I don't really see the benefit in handling the "don't write because we're using a cache" case, personally.

> Code seems like it might be messy and copy-pastey that way but might make sense to do anyway.

No worries.

Reply:
> Given that sometimes it will have to write to disk if this data is missing

This is not a common case. This should only happen if the file is deleted manually or becomes corrupt. It's a edge case we may need to worry about, but end users typically won't.

Reply:
It will have to happen once per installation.

Reply:
Which will almost certainly happen in brew update/update-report rather than `brew info` which won't even update to 4.0.0 on its own.